rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+    the data must be saved in R's custom `.rda` format.
+    We will see how to do this below.
+
+-   Manual pages go in the `man` directory.
+    The bad news is that they have to be in a sort-of-LaTeX format
+    that is only a bit less obscure than the runes inscribed on the ancient dagger
+    your colleague brought back from her latest archeological dig.
+    The good news is,
+    we can embed comments in our source code and use a tool called [roxygen2][roxygen2]
+    to extract them and format them as required.
+
+-   The `tests` directory holds the package's unit tests.
+    It should contain files with names like <code>test_<em>some_feature</em>.R</code>,
+    which should in turn contain functions named <code>test_<em>something_specific</em></code>.
+    We'll have a closer look at these in Chapter \@ref(rse-r-testing).
+
+In order to understand the rest of what follows,
+it's important to understand that R packages are distributed as compiled [byte code](glossary.html#byte-code),
+*not* as [source code](glossary.html#source-code) (which is how Python does it)
+or as [binary code](glossary.html#binary-code) (which is how app stores distribute things).
+When a package is built,
+R loads and checks the code,
+then saves the corresponding low-level instructions.
+Our R files should therefore define functions,
+not run commands immediately:
+if they do the latter,
+those commands will be executed every time the script loads,
+which is probably not what users will want.
+
+As a side effect,
+this means that if a package uses `load(something)`,
+then that `load` command is executed *while the package is being compiled*,
+and *not* while the compiled package is being loaded by a user after distribution.
+Thus,
+this simple and rather pointless "package":
+
+```{r eval=FALSE}
+library(stringr)
+
+sr <- function(text, pattern, replacement) {
+  str_replace(text, pattern, replacement)
+}
+```
+
+may not work when it's loaded by a user,
+because `stringr` may not be in memory on the user's machine at the time `str_replace` is called.
+
+How then can our packages use libraries?
+The safest way is to use [fully-qualified names](glossary.html#fully-qualified-name)
+such as `stringr::str_replace`
+every time we call a function defined somewhere outside our package.
+We will explore other options below.
+
+### Exercise: What packages do you have?
+
+What R packages are currently installed on your computer?
+How did you figure this out?
+
+### Exercise: Pros and cons
+
+What are the pros and cons of distributing packages as source code, byte code, and binary code?
+
+## How do I create a package? {#rse-r-package-create}
+
+We will create an R package called `zipffreq` (with no dashes or other special characters in its name)
+to hold word frequencies in classic English novels
+and some functions to merge and plot that data.
+The first step is to run RStudio's project creation wizard,
+select "New Project",
```

Comment:
This is definitely the easier approach, but I would suggest running:

```
usethis::create_package("path/pkgname")
```

Since this is more up to date then RStudio's builtin functionality. Plus it fixes some problems experienced later (like the `NAMESPACE` warning when running `document()`).

Reply:
FIXME added (I need to check how this works).